Multiple Set-Cookie headers in 8.4 (possibly later versions too)

**Describe the bug**

As I was deploying the docker-compose-core.yaml file inside the camunda/camunda-platform repo, I found that newer versions of connectors in version 8.4 do not properly authenticate against Operate. I got the following stacktrace and followed it to this repo:

related to compose PR https://github.com/camunda/camunda-platform/pull/616

[ see stacktrace below  in dropdown ]

Opening this in wireshark, I see that the request/response does indeed duplicate this header:

```
POST /api/login HTTP/1.1
Accept-Encoding: gzip, x-gzip, deflate
Content-Length: 27
Content-Type: application/x-www-form-urlencoded; charset=ISO-8859-1
Host: operate:8080
Connection: keep-alive
User-Agent: Apache-HttpClient/5.3.1 (Java/21.0.3)

username=demo&password=demoHTTP/1.1 204 
Vary: Origin
Vary: Access-Control-Request-Method
Vary: Access-Control-Request-Headers
Set-Cookie: OPERATE-SESSION=FBC173FD0CFF08830495B4183EA1D2E9; Path=/; HttpOnly; SameSite=Lax
Set-Cookie: OPERATE-X-CSRF-TOKEN=4ce743aa-56e8-4dfd-8fea-0cb36bd95365; Path=/; HttpOnly; SameSite=Lax
OPERATE-X-CSRF-TOKEN: 08Loutg_PHxzVga7-N1dCTCIX0apTeDQlIR3iql2nPemYugN56GNjewMXR1eYzDewPBpbVbscn7PKIH9pOcVuZ8U-M6TUd44
X-Content-Type-Options: nosniff
X-XSS-Protection: 0
Cache-Control: no-cache, no-store, max-age=0, must-revalidate
Pragma: no-cache
Expires: 0
X-Frame-Options: DENY
Content-Security-Policy: default-src 'self'; base-uri 'self'; script-src 'self'; script-src-elem 'self' cdn.jsdelivr.net; connect-src 'self' cdn.jsdelivr.net; style-src 'self' 'unsafe-inline' cdn.jsdelivr.net; img-src * data:; block-all-mixed-content; form-action 'self'; frame-ancestors 'none'; frame-src 'self' https:; object-src 'none'; font-src 'self' fonts.camunda.io cdn.jsdelivr.net; worker-src 'self' blob:; sandbox allow-forms allow-scripts allow-same-origin allow-popups
Date: Tue, 03 Sep 2024 18:01:09 GMT
Keep-Alive: timeout=60
Connection: keep-alive

```

<!-- A clear and concise description of what the bug is. -->

**To Reproduce**

1. checkout the camunda/camunda-platform repo
2. `git checkout 90ecf1933529f6134b9929f7a84bdbc4bd9ab773` (which is based on 8.4)
3. `docker compose -f docker-compose-core.yaml up -d
4. wait for `docker compose -f docker-compose-core.yaml ps` to report healthy for everything except connectors
5. get the logs for connectors.


<!--
Steps to reproduce the behavior

If possible add a minimal reproducer code sample

-->

**Expected behavior**

<!-- A clear and concise description of what you expected to happen. -->

**Log/Stacktrace**


[connectors.txt](https://github.com/user-attachments/files/16853013/connectors.txt)

<!-- If possible add the full stacktrace or log which contains the issue. -->

<details><summary>Full Stacktrace</summary>
 <p>

```
2024-09-03T18:03:00.670Z ERROR 1 --- [   scheduling-1] i.c.common.auth.SimpleAuthentication     : Authenticating for OPERATE failed due to org.apache.hc.client5.http.ClientProtocolException
2024-09-03T18:03:00.670Z ERROR 1 --- [   scheduling-1] i.c.c.r.i.i.ProcessDefinitionImporter    : Failed to import process definitions

java.lang.RuntimeException: Unable to authenticate
	at io.camunda.common.auth.SimpleAuthentication.retrieveToken(SimpleAuthentication.java:61)
	at io.camunda.common.auth.SimpleAuthentication.getTokenHeader(SimpleAuthentication.java:82)
	at io.camunda.common.http.DefaultHttpClient.retrieveToken(DefaultHttpClient.java:207)
	at io.camunda.common.http.DefaultHttpClient.post(DefaultHttpClient.java:148)
	at io.camunda.operate.CamundaOperateClient.searchProcessDefinitionResults(CamundaOperateClient.java:46)
	at io.camunda.connector.runtime.inbound.importer.ProcessDefinitionSearch.query(ProcessDefinitionSearch.java:64)
	at io.camunda.connector.runtime.inbound.importer.ProcessDefinitionImporter.scheduleImport(ProcessDefinitionImporter.java:55)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.runInternal(ScheduledMethodRunnable.java:130)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.lambda$run$2(ScheduledMethodRunnable.java:124)
	at io.micrometer.observation.Observation.observe(Observation.java:499)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:124)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source)
	at java.base/java.util.concurrent.FutureTask.runAndReset(Unknown Source)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: org.apache.hc.client5.http.ClientProtocolException: null
	at org.apache.hc.client5.http.impl.classic.CloseableHttpClient.execute(CloseableHttpClient.java:261)
	at org.apache.hc.client5.http.impl.classic.CloseableHttpClient.execute(CloseableHttpClient.java:188)
	at org.apache.hc.client5.http.impl.classic.CloseableHttpClient.execute(CloseableHttpClient.java:162)
	at io.camunda.common.auth.SimpleAuthentication.retrieveToken(SimpleAuthentication.java:41)
	... 19 common frames omitted
Caused by: org.apache.hc.core5.http.ProtocolException: multiple 'Set-Cookie' headers found
	at org.apache.hc.core5.http.message.HeaderGroup.getHeader(HeaderGroup.java:267)
	at org.apache.hc.client5.http.impl.classic.CloseableHttpResponse.getHeader(CloseableHttpResponse.java:143)
	at io.camunda.common.auth.SimpleAuthentication.lambda$retrieveToken$0(SimpleAuthentication.java:49)
	at org.apache.hc.client5.http.impl.classic.CloseableHttpClient.execute(CloseableHttpClient.java:247)
	... 22 common frames omitted

```

</p>
</details>

**Environment:**
- OS: Linux
- Deployment: SM  8.4  docker compose 
- Spring-Zeebe Version: <!-- [e.g. 8.2.4] -->
- SpringBoot Version: <!-- [e.g. 2.7.x, 3.x.x] --> 
- Configuration: <!-- [e.g. exporters etc.] -->

Connectors 8.4.8
Operate 8.4.11

--- a/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/Authentication.java
+++ b/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/Authentication.java
@@ -4,7 +4,7 @@ import java.util.Map;
 
 public interface Authentication {
 
-  Map.Entry<String, String> getTokenHeader(Product product);
+  Map<String, String> getTokenHeader(Product product);
 
   void resetToken(Product product);
 

--- a/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/DefaultNoopAuthentication.java
+++ b/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/DefaultNoopAuthentication.java
@@ -27,7 +27,7 @@ public class DefaultNoopAuthentication implements Authentication {
   }
 
   @Override
-  public Map.Entry<String, String> getTokenHeader(Product product) {
+  public Map<String, String> getTokenHeader(Product product) {
     throw new UnsupportedOperationException("Unable to determine authentication");
   }
 }

--- a/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/JwtAuthentication.java
+++ b/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/JwtAuthentication.java
@@ -1,16 +1,10 @@
 package io.camunda.common.auth;
 
-import java.lang.invoke.MethodHandles;
 import java.time.LocalDateTime;
-import java.util.AbstractMap;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.Map.Entry;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 public abstract class JwtAuthentication implements Authentication {
-  private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
 
   private final JwtConfig jwtConfig;
   private final Map<Product, JwtToken> tokens = new HashMap<>();
@@ -29,7 +23,7 @@ public abstract class JwtAuthentication implements Authentication {
   }
 
   @Override
-  public final Entry<String, String> getTokenHeader(Product product) {
+  public final Map<String, String> getTokenHeader(Product product) {
     if (!tokens.containsKey(product) || !isValid(tokens.get(product))) {
       JwtToken newToken = generateToken(product, jwtConfig.getProduct(product));
       tokens.put(product, newToken);
@@ -39,8 +33,8 @@ public abstract class JwtAuthentication implements Authentication {
 
   protected abstract JwtToken generateToken(Product product, JwtCredential credential);
 
-  private Entry<String, String> authHeader(String token) {
-    return new AbstractMap.SimpleEntry<>("Authorization", "Bearer " + token);
+  private Map<String, String> authHeader(String token) {
+    return Map.of("Authorization", "Bearer " + token);
   }
 
   private boolean isValid(JwtToken jwtToken) {

--- a/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/JwtAuthentication.java
+++ b/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/JwtAuthentication.java
@@ -1,6 +1,7 @@
 package io.camunda.common.auth;
 
 import java.time.LocalDateTime;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.Map;
 
@@ -34,7 +35,7 @@ public abstract class JwtAuthentication implements Authentication {
   protected abstract JwtToken generateToken(Product product, JwtCredential credential);
 
   private Map<String, String> authHeader(String token) {
-    return Map.of("Authorization", "Bearer " + token);
+    return Collections.singletonMap("Authorization", "Bearer " + token);
   }
 
   private boolean isValid(JwtToken jwtToken) {

--- a/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/SimpleAuthentication.java
+++ b/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/SimpleAuthentication.java
@@ -36,7 +36,7 @@ public class SimpleAuthentication implements Authentication {
       HttpPost request = buildRequest(simpleCredential);
 
       Map<String, String> headers =
-          client.execute(request, response -> List.of(response.getHeaders())).stream()
+          client.execute(request, response -> Arrays.asList(response.getHeaders())).stream()
               .reduce(
                   new HashMap<>(),
                   (map, header) -> insertHeader(map, header, product),

--- a/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/SimpleAuthentication.java
+++ b/camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/SimpleAuthentication.java
@@ -17,7 +17,7 @@ public class SimpleAuthentication implements Authentication {
   private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
 
   private final SimpleConfig simpleConfig;
-  private final Map<Product, String> tokens = new HashMap<>();
+  private final Map<Product, Map<String, String>> tokens = new HashMap<>();
 
   public SimpleAuthentication(SimpleConfig simpleConfig) {
     this.simpleConfig = simpleConfig;
@@ -31,33 +31,39 @@ public class SimpleAuthentication implements Authentication {
     return simpleConfig;
   }
 
-  private String retrieveToken(Product product, SimpleCredential simpleCredential) {
+  private Map<String, String> retrieveToken(Product product, SimpleCredential simpleCredential) {
     try (CloseableHttpClient client = HttpClients.createDefault()) {
       HttpPost request = buildRequest(simpleCredential);
-      String cookie =
-          client.execute(
-              request,
-              response -> {
-                Header[] cookieHeaders = response.getHeaders("Set-Cookie");
-                String cookieCandidate = null;
-                String cookiePrefix = product.toString().toUpperCase() + "-SESSION";
-                for (Header cookieHeader : cookieHeaders) {
-                  if (cookieHeader.getValue().startsWith(cookiePrefix)) {
-                    cookieCandidate = cookieHeader.getValue();
-                    break;
-                  }
-                }
-                return cookieCandidate;
-              });
-      if (cookie == null) {
+
+      Map<String, String> headers =
+          client.execute(request, response -> List.of(response.getHeaders())).stream()
+              .reduce(
+                  new HashMap<>(),
+                  (map, header) -> insertHeader(map, header, product),
+                  (pMap, cMap) -> {
+                    pMap.putAll(cMap);
+                    return pMap;
+                  });
+
+      if (headers.get("Cookie") == null) {
         throw new RuntimeException("Unable to authenticate due to missing Set-Cookie");
       }
-      tokens.put(product, cookie);
+      tokens.put(product, headers);
     } catch (Exception e) {
       LOG.error("Authenticating for " + product + " failed due to " + e);
       throw new RuntimeException("Unable to authenticate", e);
     }
     return tokens.get(product);
+  }
+
+  private HashMap<String, String> insertHeader(
+      HashMap<String, String> map, Header header, Product product) {
+    if (header.getValue().startsWith(product.toString().toUpperCase())) {
+      map.merge("Cookie", header.getValue(), (s, s2) -> s.concat("; ".concat(s2)));
+    } else if (header.getName().equals(product.toString().toUpperCase() + "-X-CSRF-TOKEN")) {
+      map.put(header.getName(), header.getValue());
+    }
+    return map;
   }
 
   private HttpPost buildRequest(SimpleCredential simpleCredential) {
@@ -70,16 +76,13 @@ public class SimpleAuthentication implements Authentication {
   }
 
   @Override
-  public Map.Entry<String, String> getTokenHeader(Product product) {
-    String token;
+  public Map<String, String> getTokenHeader(Product product) {
     if (tokens.containsKey(product)) {
-      token = tokens.get(product);
+      return tokens.get(product);
     } else {
       SimpleCredential simpleCredential = simpleConfig.getProduct(product);
-      token = retrieveToken(product, simpleCredential);
+      return retrieveToken(product, simpleCredential);
     }
-
-    return new AbstractMap.SimpleEntry<>("Cookie", token);
   }
 
   @Override

--- a/camunda-sdk-java/java-common/src/main/java/io/camunda/common/http/DefaultHttpClient.java
+++ b/camunda-sdk-java/java-common/src/main/java/io/camunda/common/http/DefaultHttpClient.java
@@ -10,6 +10,7 @@ import java.io.IOException;
 import java.lang.invoke.MethodHandles;
 import java.nio.charset.StandardCharsets;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicReference;
 import org.apache.hc.client5.http.classic.methods.HttpDelete;
@@ -31,14 +32,12 @@ import org.slf4j.LoggerFactory;
 public class DefaultHttpClient implements HttpClient {
 
   private static final Logger LOG = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
-
-  private String host = "";
-  private String basePath = "";
   private final Map<Product, Map<Class<?>, String>> productMap;
   private final CloseableHttpClient httpClient;
   private final Authentication authentication;
-
   private final JsonMapper jsonMapper;
+  private String host = "";
+  private String basePath = "";
 
   public DefaultHttpClient(Authentication authentication) {
     this.authentication = authentication;
@@ -78,7 +77,7 @@ public class DefaultHttpClient implements HttpClient {
   public <T> T get(Class<T> responseType, String id) {
     String url = host + basePath + retrievePath(responseType) + "/" + id;
     HttpGet httpGet = new HttpGet(url);
-    httpGet.addHeader(retrieveToken(responseType));
+    retrieveToken(responseType).forEach(httpGet::addHeader);
     T resp;
     try (CloseableHttpResponse response = httpClient.execute(httpGet)) {
       resp = parseAndRetry(response, responseType);
@@ -107,7 +106,7 @@ public class DefaultHttpClient implements HttpClient {
     }
     String url = host + basePath + resourcePath;
     HttpGet httpGet = new HttpGet(url);
-    httpGet.addHeader(retrieveToken(selector.getClass()));
+    retrieveToken(selector.getClass()).forEach(httpGet::addHeader);
     T resp;
     try (CloseableHttpResponse response = httpClient.execute(httpGet)) {
       resp = parseAndRetry(response, responseType, parameterType, selector);
@@ -128,7 +127,7 @@ public class DefaultHttpClient implements HttpClient {
   public <T> String getXml(Class<T> selector, Long key) {
     String url = host + basePath + retrievePath(selector) + "/" + key + "/xml";
     HttpGet httpGet = new HttpGet(url);
-    httpGet.addHeader(retrieveToken(selector));
+    retrieveToken(selector).forEach(httpGet::addHeader);
     String xml;
     try (CloseableHttpResponse response = httpClient.execute(httpGet)) {
       xml = parseXMLAndRetry(response, selector);
@@ -145,7 +144,7 @@ public class DefaultHttpClient implements HttpClient {
     String url = host + basePath + retrievePath(selector.getClass());
     HttpPost httpPost = new HttpPost(url);
     httpPost.addHeader("Content-Type", "application/json");
-    httpPost.addHeader(retrieveToken(selector.getClass()));
+    retrieveToken(selector.getClass()).forEach(httpPost::addHeader);
     T resp;
     String data = jsonMapper.toJson(body);
     httpPost.setEntity(new StringEntity(data));
@@ -169,7 +168,7 @@ public class DefaultHttpClient implements HttpClient {
     String resourcePath = retrievePath(selector) + "/" + key;
     String url = host + basePath + resourcePath;
     HttpDelete httpDelete = new HttpDelete(url);
-    httpDelete.addHeader(retrieveToken(selector));
+    retrieveToken(selector).forEach(httpDelete::addHeader);
     T resp = null;
     try (CloseableHttpResponse response = httpClient.execute(httpDelete)) {
       resp = parseAndRetry(response, responseType, selector);
@@ -196,7 +195,7 @@ public class DefaultHttpClient implements HttpClient {
     return path.get();
   }
 
-  private <T> Header retrieveToken(Class<T> clazz) {
+  private <T> List<? extends Header> retrieveToken(Class<T> clazz) {
     AtomicReference<Product> currentProduct = new AtomicReference<>();
     productMap.forEach(
         (product, map) -> {
@@ -204,8 +203,10 @@ public class DefaultHttpClient implements HttpClient {
             currentProduct.set(product);
           }
         });
-    Map.Entry<String, String> header = authentication.getTokenHeader(currentProduct.get());
-    return new BasicHeader(header.getKey(), header.getValue());
+    Map<String, String> tokens = authentication.getTokenHeader(currentProduct.get());
+    return tokens.entrySet().stream()
+        .map(keyValueHeader -> new BasicHeader(keyValueHeader.getKey(), keyValueHeader.getValue()))
+        .toList();
   }
 
   private <T> Product getProduct(Class<T> clazz) {

--- a/camunda-sdk-java/java-common/src/main/java/io/camunda/common/http/DefaultHttpClient.java
+++ b/camunda-sdk-java/java-common/src/main/java/io/camunda/common/http/DefaultHttpClient.java
@@ -13,6 +13,7 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
 import org.apache.hc.client5.http.classic.methods.HttpDelete;
 import org.apache.hc.client5.http.classic.methods.HttpGet;
 import org.apache.hc.client5.http.classic.methods.HttpPost;
@@ -206,7 +207,7 @@ public class DefaultHttpClient implements HttpClient {
     Map<String, String> tokens = authentication.getTokenHeader(currentProduct.get());
     return tokens.entrySet().stream()
         .map(keyValueHeader -> new BasicHeader(keyValueHeader.getKey(), keyValueHeader.getValue()))
-        .toList();
+        .collect(Collectors.toList());
   }
 
   private <T> Product getProduct(Class<T> clazz) {

--- a/camunda-sdk-java/java-common/src/test/java/io/camunda/common/http/DefaultHttpClientTest.java
+++ b/camunda-sdk-java/java-common/src/test/java/io/camunda/common/http/DefaultHttpClientTest.java
@@ -44,7 +44,7 @@ public class DefaultHttpClientTest {
     // when
     when(chClient.execute(any(HttpGet.class))).thenReturn(response);
     when(response.getCode()).thenReturn(200);
-    when(authentication.getTokenHeader(any())).thenReturn(Map.of("key", "value"));
+    when(authentication.getTokenHeader(any())).thenReturn(Collections.singletonMap("key", "value"));
     when(response.getEntity().getContent())
         .thenReturn(new ByteArrayInputStream("{\"name\" : \"test-name\"}".getBytes()));
     MyResponseClass parsedResponse = defaultHttpClient.get(MyResponseClass.class, "123");
@@ -68,7 +68,7 @@ public class DefaultHttpClientTest {
     // when
     when(chClient.execute(any(HttpPost.class))).thenReturn(response);
     when(response.getCode()).thenReturn(200);
-    when(authentication.getTokenHeader(any())).thenReturn(Map.of("key", "value"));
+    when(authentication.getTokenHeader(any())).thenReturn(Collections.singletonMap("key", "value"));
     when(response.getEntity().getContent())
         .thenReturn(new ByteArrayInputStream("[{\"name\" : \"test-name\"}]".getBytes()));
     List parsedResponse =
@@ -93,7 +93,7 @@ public class DefaultHttpClientTest {
     // when
     when(chClient.execute(any(HttpDelete.class))).thenReturn(response);
     when(response.getCode()).thenReturn(200);
-    when(authentication.getTokenHeader(any())).thenReturn(Map.of("key", "value"));
+    when(authentication.getTokenHeader(any())).thenReturn(Collections.singletonMap("key", "value"));
     when(response.getEntity().getContent())
         .thenReturn(new ByteArrayInputStream("{\"name\" : \"test-name\"}".getBytes()));
     MyResponseClass parsedResponse =

--- a/camunda-sdk-java/java-common/src/test/java/io/camunda/common/http/DefaultHttpClientTest.java
+++ b/camunda-sdk-java/java-common/src/test/java/io/camunda/common/http/DefaultHttpClientTest.java
@@ -44,8 +44,7 @@ public class DefaultHttpClientTest {
     // when
     when(chClient.execute(any(HttpGet.class))).thenReturn(response);
     when(response.getCode()).thenReturn(200);
-    when(authentication.getTokenHeader(any()))
-        .thenReturn(new AbstractMap.SimpleEntry<>("key", "value"));
+    when(authentication.getTokenHeader(any())).thenReturn(Map.of("key", "value"));
     when(response.getEntity().getContent())
         .thenReturn(new ByteArrayInputStream("{\"name\" : \"test-name\"}".getBytes()));
     MyResponseClass parsedResponse = defaultHttpClient.get(MyResponseClass.class, "123");
@@ -69,8 +68,7 @@ public class DefaultHttpClientTest {
     // when
     when(chClient.execute(any(HttpPost.class))).thenReturn(response);
     when(response.getCode()).thenReturn(200);
-    when(authentication.getTokenHeader(any()))
-        .thenReturn(new AbstractMap.SimpleEntry<>("key", "value"));
+    when(authentication.getTokenHeader(any())).thenReturn(Map.of("key", "value"));
     when(response.getEntity().getContent())
         .thenReturn(new ByteArrayInputStream("[{\"name\" : \"test-name\"}]".getBytes()));
     List parsedResponse =
@@ -95,8 +93,7 @@ public class DefaultHttpClientTest {
     // when
     when(chClient.execute(any(HttpDelete.class))).thenReturn(response);
     when(response.getCode()).thenReturn(200);
-    when(authentication.getTokenHeader(any()))
-        .thenReturn(new AbstractMap.SimpleEntry<>("key", "value"));
+    when(authentication.getTokenHeader(any())).thenReturn(Map.of("key", "value"));
     when(response.getEntity().getContent())
         .thenReturn(new ByteArrayInputStream("{\"name\" : \"test-name\"}".getBytes()));
     MyResponseClass parsedResponse =

--- a/spring-boot-starter-camunda/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfiguration.java
+++ b/spring-boot-starter-camunda/src/main/java/io/camunda/zeebe/spring/client/configuration/ZeebeClientConfiguration.java
@@ -387,10 +387,13 @@ public class ZeebeClientConfiguration implements io.camunda.zeebe.client.ZeebeCl
 
     @Override
     public void applyCredentials(Metadata headers) {
-      final Map.Entry<String, String> authHeader = authentication.getTokenHeader(Product.ZEEBE);
-      final Metadata.Key<String> authHeaderKey =
-          Metadata.Key.of(authHeader.getKey(), Metadata.ASCII_STRING_MARSHALLER);
-      headers.put(authHeaderKey, authHeader.getValue());
+      final Map<String, String> authHeader = authentication.getTokenHeader(Product.ZEEBE);
+      authHeader.forEach(
+          (key, value) -> {
+            final Metadata.Key<String> authHeaderKey =
+                Metadata.Key.of(key, Metadata.ASCII_STRING_MARSHALLER);
+            headers.put(authHeaderKey, value);
+          });
     }
 
     @Override

--- a/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationOidcTest.java
+++ b/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationOidcTest.java
@@ -11,7 +11,7 @@ import io.camunda.common.auth.Product;
 import io.camunda.common.auth.SelfManagedAuthentication;
 import io.camunda.zeebe.spring.client.configuration.AuthenticationConfiguration;
 import java.time.Instant;
-import java.util.Map;
+import java.util.Collections;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
@@ -49,7 +49,7 @@ public class AuthenticationConfigurationOidcTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.OPERATE))
         .isNotNull()
-        .isEqualTo(Map.of("Authorization", "Bearer " + accessToken));
+        .isEqualTo(Collections.singletonMap("Authorization", "Bearer " + accessToken));
     verify(
         postRequestedFor(urlEqualTo("/auth-server/protocol/openid-connect/token"))
             .withHeader("Content-Type", equalTo("application/x-www-form-urlencoded")));
@@ -68,7 +68,7 @@ public class AuthenticationConfigurationOidcTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.TASKLIST))
         .isNotNull()
-        .isEqualTo(Map.of("Authorization", "Bearer " + accessToken));
+        .isEqualTo(Collections.singletonMap("Authorization", "Bearer " + accessToken));
     verify(
         postRequestedFor(urlEqualTo("/auth-server/protocol/openid-connect/token"))
             .withHeader("Content-Type", equalTo("application/x-www-form-urlencoded")));
@@ -87,7 +87,7 @@ public class AuthenticationConfigurationOidcTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.OPTIMIZE))
         .isNotNull()
-        .isEqualTo(Map.of("Authorization", "Bearer " + accessToken));
+        .isEqualTo(Collections.singletonMap("Authorization", "Bearer " + accessToken));
     verify(
         postRequestedFor(urlEqualTo("/auth-server/protocol/openid-connect/token"))
             .withHeader("Content-Type", equalTo("application/x-www-form-urlencoded")));
@@ -106,7 +106,7 @@ public class AuthenticationConfigurationOidcTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.ZEEBE))
         .isNotNull()
-        .isEqualTo(Map.of("Authorization", "Bearer " + accessToken));
+        .isEqualTo(Collections.singletonMap("Authorization", "Bearer " + accessToken));
     verify(
         postRequestedFor(urlEqualTo("/auth-server/protocol/openid-connect/token"))
             .withHeader("Content-Type", equalTo("application/x-www-form-urlencoded")));

--- a/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationOidcTest.java
+++ b/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationOidcTest.java
@@ -11,6 +11,7 @@ import io.camunda.common.auth.Product;
 import io.camunda.common.auth.SelfManagedAuthentication;
 import io.camunda.zeebe.spring.client.configuration.AuthenticationConfiguration;
 import java.time.Instant;
+import java.util.Map;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
@@ -48,7 +49,7 @@ public class AuthenticationConfigurationOidcTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.OPERATE))
         .isNotNull()
-        .isEqualTo(entry("Authorization", "Bearer " + accessToken));
+        .isEqualTo(Map.of("Authorization", "Bearer " + accessToken));
     verify(
         postRequestedFor(urlEqualTo("/auth-server/protocol/openid-connect/token"))
             .withHeader("Content-Type", equalTo("application/x-www-form-urlencoded")));
@@ -67,7 +68,7 @@ public class AuthenticationConfigurationOidcTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.TASKLIST))
         .isNotNull()
-        .isEqualTo(entry("Authorization", "Bearer " + accessToken));
+        .isEqualTo(Map.of("Authorization", "Bearer " + accessToken));
     verify(
         postRequestedFor(urlEqualTo("/auth-server/protocol/openid-connect/token"))
             .withHeader("Content-Type", equalTo("application/x-www-form-urlencoded")));
@@ -86,7 +87,7 @@ public class AuthenticationConfigurationOidcTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.OPTIMIZE))
         .isNotNull()
-        .isEqualTo(entry("Authorization", "Bearer " + accessToken));
+        .isEqualTo(Map.of("Authorization", "Bearer " + accessToken));
     verify(
         postRequestedFor(urlEqualTo("/auth-server/protocol/openid-connect/token"))
             .withHeader("Content-Type", equalTo("application/x-www-form-urlencoded")));
@@ -105,7 +106,7 @@ public class AuthenticationConfigurationOidcTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.ZEEBE))
         .isNotNull()
-        .isEqualTo(entry("Authorization", "Bearer " + accessToken));
+        .isEqualTo(Map.of("Authorization", "Bearer " + accessToken));
     verify(
         postRequestedFor(urlEqualTo("/auth-server/protocol/openid-connect/token"))
             .withHeader("Content-Type", equalTo("application/x-www-form-urlencoded")));

--- a/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationSaasTest.java
+++ b/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationSaasTest.java
@@ -8,6 +8,7 @@ import io.camunda.common.auth.Authentication;
 import io.camunda.common.auth.Product;
 import io.camunda.common.auth.SaaSAuthentication;
 import io.camunda.zeebe.spring.client.configuration.AuthenticationConfiguration;
+import java.util.Map;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
@@ -46,7 +47,7 @@ public class AuthenticationConfigurationSaasTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.OPERATE))
         .isNotNull()
-        .isEqualTo(entry("Authorization", "Bearer " + ACCESS_TOKEN));
+        .isEqualTo(Map.of("Authorization", "Bearer " + ACCESS_TOKEN));
     verify(
         postRequestedFor(urlEqualTo("/auth-server"))
             .withHeader("Content-Type", equalTo("application/json")));
@@ -64,7 +65,7 @@ public class AuthenticationConfigurationSaasTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.TASKLIST))
         .isNotNull()
-        .isEqualTo(entry("Authorization", "Bearer " + ACCESS_TOKEN));
+        .isEqualTo(Map.of("Authorization", "Bearer " + ACCESS_TOKEN));
     verify(
         postRequestedFor(urlEqualTo("/auth-server"))
             .withHeader("Content-Type", equalTo("application/json")));
@@ -82,7 +83,7 @@ public class AuthenticationConfigurationSaasTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.OPTIMIZE))
         .isNotNull()
-        .isEqualTo(entry("Authorization", "Bearer " + ACCESS_TOKEN));
+        .isEqualTo(Map.of("Authorization", "Bearer " + ACCESS_TOKEN));
     verify(
         postRequestedFor(urlEqualTo("/auth-server"))
             .withHeader("Content-Type", equalTo("application/json")));
@@ -100,7 +101,7 @@ public class AuthenticationConfigurationSaasTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.ZEEBE))
         .isNotNull()
-        .isEqualTo(entry("Authorization", "Bearer " + ACCESS_TOKEN));
+        .isEqualTo(Map.of("Authorization", "Bearer " + ACCESS_TOKEN));
     verify(
         postRequestedFor(urlEqualTo("/auth-server"))
             .withHeader("Content-Type", equalTo("application/json")));

--- a/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationSaasTest.java
+++ b/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationSaasTest.java
@@ -8,7 +8,7 @@ import io.camunda.common.auth.Authentication;
 import io.camunda.common.auth.Product;
 import io.camunda.common.auth.SaaSAuthentication;
 import io.camunda.zeebe.spring.client.configuration.AuthenticationConfiguration;
-import java.util.Map;
+import java.util.Collections;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
@@ -47,7 +47,7 @@ public class AuthenticationConfigurationSaasTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.OPERATE))
         .isNotNull()
-        .isEqualTo(Map.of("Authorization", "Bearer " + ACCESS_TOKEN));
+        .isEqualTo(Collections.singletonMap("Authorization", "Bearer " + ACCESS_TOKEN));
     verify(
         postRequestedFor(urlEqualTo("/auth-server"))
             .withHeader("Content-Type", equalTo("application/json")));
@@ -65,7 +65,7 @@ public class AuthenticationConfigurationSaasTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.TASKLIST))
         .isNotNull()
-        .isEqualTo(Map.of("Authorization", "Bearer " + ACCESS_TOKEN));
+        .isEqualTo(Collections.singletonMap("Authorization", "Bearer " + ACCESS_TOKEN));
     verify(
         postRequestedFor(urlEqualTo("/auth-server"))
             .withHeader("Content-Type", equalTo("application/json")));
@@ -83,7 +83,7 @@ public class AuthenticationConfigurationSaasTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.OPTIMIZE))
         .isNotNull()
-        .isEqualTo(Map.of("Authorization", "Bearer " + ACCESS_TOKEN));
+        .isEqualTo(Collections.singletonMap("Authorization", "Bearer " + ACCESS_TOKEN));
     verify(
         postRequestedFor(urlEqualTo("/auth-server"))
             .withHeader("Content-Type", equalTo("application/json")));
@@ -101,7 +101,7 @@ public class AuthenticationConfigurationSaasTest {
                             .put("expires_in", 300))));
     assertThat(authentication.getTokenHeader(Product.ZEEBE))
         .isNotNull()
-        .isEqualTo(Map.of("Authorization", "Bearer " + ACCESS_TOKEN));
+        .isEqualTo(Collections.singletonMap("Authorization", "Bearer " + ACCESS_TOKEN));
     verify(
         postRequestedFor(urlEqualTo("/auth-server"))
             .withHeader("Content-Type", equalTo("application/json")));

--- a/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationSimpleTest.java
+++ b/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationSimpleTest.java
@@ -8,6 +8,8 @@ import io.camunda.common.auth.Authentication;
 import io.camunda.common.auth.Product;
 import io.camunda.common.auth.SimpleAuthentication;
 import io.camunda.zeebe.spring.client.configuration.AuthenticationConfiguration;
+import java.util.Map;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.test.context.SpringBootTest;
@@ -28,6 +30,11 @@ public class AuthenticationConfigurationSimpleTest {
     assertThat(authentication).isExactlyInstanceOf(SimpleAuthentication.class);
   }
 
+  @BeforeEach
+  void setUp() {
+    authentication.resetToken(Product.OPERATE);
+  }
+
   @Test
   void shouldHaveOperateAuth() {
     stubFor(
@@ -39,7 +46,10 @@ public class AuthenticationConfigurationSimpleTest {
                         "OPERATE-X-CSRF-TOKEN=139196d4-7768-451c-aa66-078e1ed74785")));
     assertThat(authentication.getTokenHeader(Product.OPERATE))
         .isNotNull()
-        .isEqualTo(entry("Cookie", "OPERATE-SESSION=3205A03818447100591792E774DB8AF6"));
+        .isEqualTo(
+            Map.of(
+                "Cookie",
+                "OPERATE-SESSION=3205A03818447100591792E774DB8AF6; OPERATE-X-CSRF-TOKEN=139196d4-7768-451c-aa66-078e1ed74785"));
     verify(
         postRequestedFor(urlEqualTo("/api/login"))
             .withHeader(
@@ -57,7 +67,32 @@ public class AuthenticationConfigurationSimpleTest {
                         "OPERATE-X-CSRF-TOKEN=139196d4-7768-451c-aa66-078e1ed74785")));
     assertThat(authentication.getTokenHeader(Product.TASKLIST))
         .isNotNull()
-        .isEqualTo(entry("Cookie", "TASKLIST-SESSION=3205A03818447100591792E774DB8AF6"));
+        .isEqualTo(Map.of("Cookie", "TASKLIST-SESSION=3205A03818447100591792E774DB8AF6"));
+    verify(
+        postRequestedFor(urlEqualTo("/api/login"))
+            .withHeader(
+                "Content-Type", equalTo("application/x-www-form-urlencoded; charset=ISO-8859-1")));
+  }
+
+  @Test
+  void shouldHaveCSRFToken() {
+    stubFor(
+        post("/api/login")
+            .willReturn(
+                ok().withHeader(
+                        "Set-Cookie", "OPERATE-X-CSRF-TOKEN=139196d4-7768-451c-aa66-078e1ed74785")
+                    .withHeader(
+                        "OPERATE-X-CSRF-TOKEN",
+                        "WwbfQ33kHNEHu99ioC39yCMuVE2JjQnK_vYEpGPQGxBv1Nfn10pSDzapEpzfcJJvbiE2kG6jDMbFVVBSMy9K_K5dlV1")));
+
+    assertThat(authentication.getTokenHeader(Product.OPERATE))
+        .isNotNull()
+        .isEqualTo(
+            Map.of(
+                "Cookie",
+                "OPERATE-X-CSRF-TOKEN=139196d4-7768-451c-aa66-078e1ed74785",
+                "OPERATE-X-CSRF-TOKEN",
+                "WwbfQ33kHNEHu99ioC39yCMuVE2JjQnK_vYEpGPQGxBv1Nfn10pSDzapEpzfcJJvbiE2kG6jDMbFVVBSMy9K_K5dlV1"));
     verify(
         postRequestedFor(urlEqualTo("/api/login"))
             .withHeader(

--- a/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationSimpleTest.java
+++ b/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/AuthenticationConfigurationSimpleTest.java
@@ -8,6 +8,8 @@ import io.camunda.common.auth.Authentication;
 import io.camunda.common.auth.Product;
 import io.camunda.common.auth.SimpleAuthentication;
 import io.camunda.zeebe.spring.client.configuration.AuthenticationConfiguration;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.Map;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
@@ -47,7 +49,7 @@ public class AuthenticationConfigurationSimpleTest {
     assertThat(authentication.getTokenHeader(Product.OPERATE))
         .isNotNull()
         .isEqualTo(
-            Map.of(
+            Collections.singletonMap(
                 "Cookie",
                 "OPERATE-SESSION=3205A03818447100591792E774DB8AF6; OPERATE-X-CSRF-TOKEN=139196d4-7768-451c-aa66-078e1ed74785"));
     verify(
@@ -67,7 +69,9 @@ public class AuthenticationConfigurationSimpleTest {
                         "OPERATE-X-CSRF-TOKEN=139196d4-7768-451c-aa66-078e1ed74785")));
     assertThat(authentication.getTokenHeader(Product.TASKLIST))
         .isNotNull()
-        .isEqualTo(Map.of("Cookie", "TASKLIST-SESSION=3205A03818447100591792E774DB8AF6"));
+        .isEqualTo(
+            Collections.singletonMap(
+                "Cookie", "TASKLIST-SESSION=3205A03818447100591792E774DB8AF6"));
     verify(
         postRequestedFor(urlEqualTo("/api/login"))
             .withHeader(
@@ -88,7 +92,7 @@ public class AuthenticationConfigurationSimpleTest {
     assertThat(authentication.getTokenHeader(Product.OPERATE))
         .isNotNull()
         .isEqualTo(
-            Map.of(
+            mapOf(
                 "Cookie",
                 "OPERATE-X-CSRF-TOKEN=139196d4-7768-451c-aa66-078e1ed74785",
                 "OPERATE-X-CSRF-TOKEN",
@@ -97,5 +101,12 @@ public class AuthenticationConfigurationSimpleTest {
         postRequestedFor(urlEqualTo("/api/login"))
             .withHeader(
                 "Content-Type", equalTo("application/x-www-form-urlencoded; charset=ISO-8859-1")));
+  }
+
+  private static Map<String, String> mapOf(String key1, String value1, String key2, String value2) {
+    Map<String, String> map = new HashMap<>();
+    map.put(key1, value1);
+    map.put(key2, value2);
+    return map;
   }
 }

--- a/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/ZeebeClientConfigurationTest.java
+++ b/spring-boot-starter-camunda/src/test/java/io/camunda/zeebe/spring/client/config/ZeebeClientConfigurationTest.java
@@ -14,7 +14,7 @@ import io.camunda.zeebe.spring.client.properties.ZeebeClientConfigurationPropert
 import io.grpc.ClientInterceptor;
 import java.util.Collections;
 import java.util.List;
-import java.util.Map.Entry;
+import java.util.Map;
 import org.junit.jupiter.api.Test;
 import org.springframework.mock.env.MockEnvironment;
 
@@ -41,7 +41,7 @@ public class ZeebeClientConfigurationTest {
   private static Authentication authentication() {
     return new Authentication() {
       @Override
-      public Entry<String, String> getTokenHeader(Product product) {
+      public Map<String, String> getTokenHeader(Product product) {
         return null;
       }
 

